In [80]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [105]:
corona = pd.read_csv('datasets/2019_nCoV_20200121_20200206.csv')
corona.sample(10)

Province/State  Country/Region      Last Update  Confirmed  Suspected  \
700            Hebei  Mainland China    2/1/2020 1:52       96.0        NaN   
1847             NaN           Japan  1/22/2020 12:00        2.0        NaN   
263   San Benito, CA   United States      2/3/20 3:53        2.0        NaN   
1337       Hong Kong       Hong Kong  1/27/2020 20:30        8.0        NaN   
434     Heilongjiang  Mainland China    2/2/2020 3:23       95.0        NaN   
245              NaN         Germany     2/3/20 20:53       12.0        NaN   
910            Tibet  Mainland China  1/30/2020 21:30        1.0        NaN   
1437          Shanxi  Mainland China   1/27/2020 9:00       13.0        NaN   
233          Guizhou  Mainland China     2/4/20 10:43       58.0        NaN   
823         Shandong  Mainland China  1/31/2020 14:00      184.0        NaN   

      Recovered  Death  
700         0.0    1.0  
1847        NaN    NaN  
263         0.0    0.0  
1337        NaN    NaN  
434         2.0    2.0  
245         0.0    0.0  
910         NaN    NaN  
1437        NaN    NaN  
233         2.0    0.0  
823         2.0    NaN

In [106]:
def data_inv(df):
    print('Number of Rows: ', df.shape[0])
    print('Number of Columns: ', df.shape[1])
    print('-' * 30)
    print('Dataset Columns: \n')
    print(df.columns)
    print('-' * 30)
    print('Datatype of Each Column: \n')
    print(df.dtypes)
    print('-' * 30)
    print('Missing Rows in Each Column: \n')
    c=df.isnull().sum()
    print(c[c > 0])
data_inv(corona)

Number of Rows:  1877
Number of Columns:  7
------------------------------
Dataset Columns: 

Index(['Province/State', 'Country/Region', 'Last Update', 'Confirmed',
       'Suspected', 'Recovered', 'Death'],
      dtype='object')
------------------------------
Datatype of Each Column: 

Province/State     object
Country/Region     object
Last Update        object
Confirmed         float64
Suspected         float64
Recovered         float64
Death             float64
dtype: object
------------------------------
Missing Rows in Each Column: 

Province/State     459
Confirmed           30
Suspected         1789
Recovered          876
Death              999
dtype: int64


## Dropping Columns
- 'Suspected' has 95% missing values (1789/1877)

In [107]:
corona.drop('Suspected', inplace=True, axis=1)
corona.head()

Province/State  Country/Region   Last Update  Confirmed  Recovered  Death
0          Hubei  Mainland China  2/5/20 16:43    16678.0      538.0  479.0
1      Guangdong  Mainland China  2/5/20 13:23      895.0       49.0    0.0
2       Zhejiang  Mainland China  2/5/20 15:13      895.0       78.0    0.0
3          Henan  Mainland China  2/5/20 15:03      764.0       47.0    2.0
4          Hunan  Mainland China  2/5/20 15:23      661.0       54.0    0.0

## Change the Index
- No fields are unique, so we'll stick with the provided index. This may need to be remade in case any rows need removal later. 

In [108]:
for column in corona.columns:
    if corona[column].is_unique is True:
        print(f"{column} is unique")
    else:
        print(f"{column} is NOT unique")

Province/State is NOT unique
Country/Region is NOT unique
Last Update is NOT unique
Confirmed is NOT unique
Recovered is NOT unique
Death is NOT unique


## Clean-up Fields
- 'Last Update' will need to be in datetime
- 'Mainland China' in 'Country/Region' needs to be changed to just 'China'

In [109]:
corona['Country/Region'].value_counts()

Mainland China          1028
United States            154
Australia                 81
Canada                    50
South Korea               34
Taiwan                    34
Thailand                  34
Hong Kong                 34
Japan                     34
Macau                     33
Singapore                 32
Vietnam                   32
Malaysia                  30
France                    30
Nepal                     29
Cambodia                  24
Sri Lanka                 23
Germany                   22
United Arab Emirates      18
Finland                   17
Philippines               16
India                     15
Italy                     14
Sweden                    13
UK                        13
Russia                    13
Spain                     12
Belgium                    3
Ivory Coast                2
Colombia                   1
Brazil                     1
Mexico                     1
Name: Country/Region, dtype: int64

In [110]:
corona.replace(to_replace='Mainland China', value='China', inplace=True)
corona['Country/Region'].value_counts().head()

China            1028
United States     154
Australia          81
Canada             50
Japan              34
Name: Country/Region, dtype: int64

In [111]:
corona['Country/Region'].nunique()

32

In [112]:
corona['Confirmed'].fillna(0, inplace=True)
corona['Recovered'].fillna(0, inplace=True)
corona['Death'].fillna(0, inplace=True)
corona['Province/State'].fillna('Unknown', inplace=True)
corona.sample(5)

Province/State Country/Region      Last Update  Confirmed  Recovered  \
1098        Unknown      Sri Lanka  1/29/2020 14:30        1.0        0.0   
1494        Qinghai          China  1/26/2020 23:00        4.0        0.0   
104         Unknown    South Korea     2/4/20 15:33       16.0        0.0   
1197        Unknown        Vietnam  1/28/2020 23:00        2.0        0.0   
138     Seattle, WA  United States     2/1/20 19:43        1.0        0.0   

      Death  
1098    0.0  
1494    0.0  
104     0.0  
1197    0.0  
138     0.0

In [113]:
corona['Last Update'] = pd.to_datetime(corona['Last Update'])
corona.sample(10)

Province/State        Country/Region         Last Update  Confirmed  \
457         Unknown  United Arab Emirates 2020-02-02 05:43:00        5.0   
1464       Zhejiang                 China 2020-01-26 23:00:00      128.0   
1012         Hainan                 China 2020-01-29 21:00:00       43.0   
1261      Guangdong                 China 2020-01-28 13:00:00      207.0   
1328          Hebei                 China 2020-01-27 20:30:00       33.0   
442         Ningxia                 China 2020-02-02 01:43:00       28.0   
185         Unknown                 India 2020-02-03 21:43:00        3.0   
1115       Shanghai                 China 2020-01-29 13:30:00       96.0   
1778      Hong Kong             Hong Kong 2020-01-23 12:00:00        2.0   
1604        Beijing                 China 2020-01-25 12:00:00       41.0   

      Recovered  Death  
457         0.0    0.0  
1464        1.0    0.0  
1012        0.0    1.0  
1261        4.0    0.0  
1328        0.0    1.0  
442         0.0    0.0  
185         0.0    0.0  
1115        4.0    1.0  
1778        0.0    0.0  
1604        2.0    0.0

In [114]:
corona.dtypes

Province/State            object
Country/Region            object
Last Update       datetime64[ns]
Confirmed                float64
Recovered                float64
Death                    float64
dtype: object

## Renaming Columns and Skipping Rows
- 'Province/State' -> 'Division'
- 'Country/Region' -> 'Country'
- 'Last Update' -> 'Date'

In [115]:
new_names = {'Province/State': 'Region',
             'Country/Region': 'Country',
             'Last Update': 'Date'}
corona.rename(columns=new_names, inplace=True)
corona.sample(5)

Region    Country                Date  Confirmed  Recovered  \
1205  New South Wales  Australia 2020-01-28 23:00:00        4.0        0.0   
893            Yunnan      China 2020-01-30 21:30:00       76.0        0.0   
520             Macau      Macau 2020-02-02 04:23:00        8.0        0.0   
1834          Qinghai      China 2020-01-22 12:00:00        0.0        0.0   
1388        Hong Kong  Hong Kong 2020-01-27 19:00:00        8.0        0.0   

      Death  
1205    0.0  
893     0.0  
520     0.0  
1834    0.0  
1388    0.0

### EDA

In [117]:
corona['Death_Pct'] = corona['Death'] / corona['Confirmed']
corona.head(3)

Region    Country                Date  Confirmed  Recovered  \
725   New South Wales  Australia 2020-02-01 18:12:00        4.0        2.0   
1684          Unknown  Australia 2020-01-25 00:00:00        1.0        0.0   
1099  New South Wales  Australia 2020-01-29 14:30:00        4.0        0.0   

      Death  Death_Pct  
725     0.0        0.0  
1684    0.0        0.0  
1099    0.0        0.0

In [118]:
corona['Recoverd_Pct'] = corona['Recovered'] / corona['Confirmed']
corona.head(3)

Region    Country                Date  Confirmed  Recovered  \
725   New South Wales  Australia 2020-02-01 18:12:00        4.0        2.0   
1684          Unknown  Australia 2020-01-25 00:00:00        1.0        0.0   
1099  New South Wales  Australia 2020-01-29 14:30:00        4.0        0.0   

      Death  Death_Pct  Recoverd_Pct  
725     0.0        0.0           0.5  
1684    0.0        0.0           0.0  
1099    0.0        0.0           0.0

In [119]:
sum_row = corona[['Confirmed', 'Recovered', 'Death']].sum()
sum_row

Confirmed    305891.0
Recovered      8634.0
Death          6553.0
dtype: float64

Something isn't right here. There isn't 6,553 deaths associated with Coronavirus yet. I wonder if the data is showing cumulative data per date or location.

In [120]:
aus_df = df[corona['Country']=='Australia']
aus_df.head()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


Province/State Country/Region   Last Update  Confirmed  Suspected  \
42   New South Wales      Australia  2/1/20 18:12        4.0        NaN   
43          Victoria      Australia  2/1/20 18:12        4.0        NaN   
44        Queensland      Australia  2/4/20 16:53        3.0        NaN   
46   South Australia      Australia  2/2/20 22:33        2.0        NaN   
112  New South Wales      Australia  2/1/20 18:12        4.0        NaN   

     Recovered  Death  
42         2.0    0.0  
43         0.0    0.0  
44         0.0    0.0  
46         0.0    0.0  
112        2.0    0.0

In [121]:
aus_df.describe()

Confirmed  Suspected  Recovered  Death
count  80.000000        1.0  48.000000   44.0
mean    2.975000        1.0   0.625000    0.0
std     1.168955        NaN   0.936835    0.0
min     1.000000        1.0   0.000000    0.0
25%     2.000000        1.0   0.000000    0.0
50%     3.500000        1.0   0.000000    0.0
75%     4.000000        1.0   2.000000    0.0
max     5.000000        1.0   2.000000    0.0

In [122]:
aus_sum_row = aus_df[['Confirmed', 'Recovered', 'Death']].sum()
aus_sum_row

Confirmed    238.0
Recovered     30.0
Death          0.0
dtype: float64

In [138]:
aus_df.sort_values(by=['Province/State','Last Update']).head()

Province/State Country/Region         Last Update  Confirmed  \
1412  New South Wales      Australia 2020-01-27 19:00:00        4.0   
1360  New South Wales      Australia 2020-01-27 20:30:00        4.0   
1308  New South Wales      Australia 2020-01-28 13:00:00        4.0   
1257  New South Wales      Australia 2020-01-28 18:00:00        4.0   
1205  New South Wales      Australia 2020-01-28 23:00:00        4.0   

      Suspected  Recovered  Death  
1412        NaN        NaN    NaN  
1360        NaN        NaN    NaN  
1308        NaN        NaN    NaN  
1257        NaN        NaN    NaN  
1205        NaN        NaN    NaN

In [125]:
aus_df['Confirmed'].value_counts()

4.0    39
2.0    21
1.0    11
3.0     8
5.0     1
Name: Confirmed, dtype: int64

In [135]:
aus_df['Last Update'] = pd.to_datetime(aus_df['Last Update'])
aus_df.dtypes

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Province/State            object
Country/Region            object
Last Update       datetime64[ns]
Confirmed                float64
Suspected                float64
Recovered                float64
Death                    float64
dtype: object

In [140]:
aus_df_sub = aus_df[['Province/State', 
                     'Confirmed', 
                     'Recovered', 
                     'Death']].groupby('Province/State').sum()
aus_df_sub

Confirmed  Recovered  Death
Province/State                              
New South Wales       92.0       30.0    0.0
Queensland            38.0        0.0    0.0
South Australia       22.0        0.0    0.0
Victoria              64.0        0.0    0.0

In [141]:
aus_df_sub = aus_df[['Province/State', 
                     'Confirmed', 
                     'Recovered', 
                     'Death']].groupby('Province/State').count()
aus_df_sub

Confirmed  Recovered  Death
Province/State                              
New South Wales         23         15     11
Queensland              16         11     11
South Australia         12         11     11
Victoria                23         11     11

In [142]:
aus_df_sub = aus_df[['Province/State', 
                     'Confirmed', 
                     'Recovered', 
                     'Death']].groupby('Province/State').max()
aus_df_sub

Confirmed  Recovered  Death
Province/State                              
New South Wales        4.0        2.0    0.0
Queensland             3.0        0.0    0.0
South Australia        2.0        0.0    0.0
Victoria               4.0        0.0    0.0

If we divide the 'Confirmed' sum by the 'Confirmed' count for **New South Wales** we get exactly 4.0 'Confirmed' for every entry. This number also happens to be the max. The same is True for 'Recovered'. 

I believe this dataset is marking cumulative numbers for every location. Therefore, we should be able to disregard all rows that are not the latest count for each location. 